### ISSUE TITLE
KPU run_yolo2 from SD card image

### ISSUE THREAD

**Issue by wizath** [sipeed/MaixPy-v1#110]
I want to try KPU with image sourced from SD card. But when I load image
```
>>> x = image.Image('/sd/1.jpg')
>>> code = kpu.run_yolo2(task, x)
[MAIXPY]kpu: pix_ai is NULL!
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: [MAIXPY]kpu: check img format err!
>>> x
{"w":320, "h":240, "type"="rgb565", "size":153600}
```

Anyway, when I save image from sensor.snapshot(), then load it, the error is exactly the same
```
>>> img = sensor.snapshot()
>>> img.save('/sd/1.jpg')
{"w":320, "h":240, "type"="rgb565", "size":153600}
>>> x = image.Image('/sd/1.jpg')
>>> code = kpu.run_yolo2(task, x)
[MAIXPY]kpu: pix_ai is NULL!
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: [MAIXPY]kpu: check img format err!
```

Does anyone have any hint what could be wrong?

**Comment by Vinschni** [comment]
I tried to generate images / mjpeg data with ffmpeg as input for kpu yolo2 but were not successful.
`mediainfo` of a avi recorded to SD gives me:
```
Complete name                            : capture.avi
Format                                   : AVI
Format/Info                              : Audio Video Interleave
File size                                : 470 KiB

Video
ID                                       : 0
Format                                   : JPEG
Codec ID                                 : MJPG
Width                                    : 320 pixels
Height                                   : 240 pixels
Display aspect ratio                     : 4:3
Frame rate                               : 4.000 FPS
Color space                              : YUV
Chroma subsampling                       : 4:2:2
Bit depth                                : 8 bits
Compression mode                         : Lossy
```

I have the same error, I'm unable to generate valid input data for yolo2.
The example video provided here: https://maixpy.sipeed.com/en/libs/machine_vision/video.html#example-1-play-avi-video
can be played, but not used to for classification. 

I used the algorithm from demo (https://github.com/sipeed/MaixPy/blob/ea8cd88cf25a909e94aa97e1379679c0cde67a3f/ports/k210-freertos/docs/demo.md#video)
```
import os, sys, time, image, lcd

f = open("badapple.avi", "rb")
index = 0
tmp = f.read(4)
while tmp !=b'movi':
    index += 1
    a = f.seek(index,0)
    tmp = f.read(4)
index += 4
movi_location = index
f.seek(0,2)
f_size = f.tell()
index = f_size
while tmp != b'idx1':
    index -= 1
    a = f.seek(index, 0)
    tmp = f.read(4)
count = 0
lcd.init()
tim = time.ticks_ms()
while tmp:
    while tmp != b'00dc':
        index += 1
        a = f.seek(index, 0)
        tmp = f.read(4)
    index += 8       #dwflag +4bytes from dwflag--->dwoffset=frame_loaction
    a = f.seek(index,0)
    tmp = f.read(4)
    frame_location = movi_location + int.from_bytes(tmp, "little")
    a = f.seek(frame_location,0)
    tmp = f.read(4)
    frame_size = int.from_bytes(tmp, 'little')
    frame_location += 4 #//only data length removed
    a = f.seek(frame_location, 0)
    img = image.Image(f, copy_to_fb=1)
    a = img.draw_string(2,2,"%.2ffps" %(1000/(time.ticks_ms()-tim)), color=(255,0,0), scale=2)
    lcd.display(img)
    a = f.seek(index, 0)
    count += 1
    tim = time.ticks_ms()

f.close()
```

to get the video frame by frame as input for yolo2.

Error is the same `ValueError: [MAIXPY]kpu: check img format err!`

**Comment by wizath** [comment]
As I was searching the source code for any hint, it seems to be quite reasonable.

We have image_t struct with this weird pix_ai field.
https://github.com/sipeed/MaixPy/blob/a21a2ba1c305252543d43a512ace708a9413ac00/ports/k210-freertos/mpy_support/omv/img/include/imlib.h#L418

Then in Maix_kpu there is this check
https://github.com/sipeed/MaixPy/blob/a21a2ba1c305252543d43a512ace708a9413ac00/ports/k210-freertos/mpy_support/Maix/Maix_kpu.c#L70

TL;DR Loading images from file will read image data only to _data field_, not the _pix_ai_ and that is the cause of that validation error.

But now how to copy image contents (and which format) to the pix_ai field of image struct? According to comment this is for _speedup_.

**Comment by Vinschni** [comment]
I found a function which probably does what you need:
https://github.com/sipeed/MaixPy/blob/f075d9f8a336829826dd4e6eb05d9a43a0dbf689/ports/k210-freertos/mpy_support/omv/py/py_image.c#L5577

call it like `img.pix_to_ai()` to use it.

**Comment by wizath** [comment]
Thanks! Overlooked it, it now works fine, except some heap issues :champagne:

**Comment by DeqingSun** [comment]
> I found a function which probably does what you need:
> https://github.com/sipeed/MaixPy/blob/f075d9f8a336829826dd4e6eb05d9a43a0dbf689/ports/k210-freertos/mpy_support/omv/py/py_image.c#L5577
> 
> call it like `img.pix_to_ai()` to use it.

Thanks, this helped me as well. Although this may cause heap memory leak, using ```del img``` can release the heap memory.

**Comment by lastrei** [comment]
> > I found a function which probably does what you need:
> > https://github.com/sipeed/MaixPy/blob/f075d9f8a336829826dd4e6eb05d9a43a0dbf689/ports/k210-freertos/mpy_support/omv/py/py_image.c#L5577
> > 
> > call it like `img.pix_to_ai()` to use it.
> 
> Thanks, this helped me as well. Although this may cause heap memory leak, using `del img` can release the heap memory.

but when i use pix_to_ai still return none?

may i see your code ?

**Comment by wiebevr** [comment]
FWIW `pix_to_ai()` works in place. The following works for me:
```
img = image.Image('/flash/image.jpg').to_rgb565(copy=True)
img.pix_to_ai()
```
